packages/runtime/src/runtime/hooks/useEnsembleAction.tsx

```diff
@@ -120,6 +131,163 @@ export const useInvokeApi: EnsembleActionHook<InvokeAPIAction> = (action) => {
   return invokeApi;
 };
 
+export const usePickFiles: EnsembleActionHook<PickFilesAction> = (
+  action?: PickFilesAction,
+) => {
+  const [files, setFiles] = useState<FileList>();
+  const onCompleteAction = useEnsembleAction(action?.onComplete);
+
+  useRegisterBindings(
+    {
+      files,
+    },
+    action?.id,
+    {
+      setFiles,
+    },
+  );
+
+  const inputEl = useMemo(() => {
+    const input = document.createElement("input");
+    input.type = "file";
+    input.multiple = action?.allowMultiple || false;
+    input.accept =
+      action?.allowedExtensions?.map((ext) => ".".concat(ext))?.toString() ||
+      "*/*";
+
+    input.onchange = (event: Event): void => {
+      const selectedFiles =
+        (event.target as HTMLInputElement).files || undefined;
+
+      if (selectedFiles) setFiles(selectedFiles);
+
+      if (onCompleteAction) onCompleteAction?.callback();
+    };
+
+    return input;
+  }, [action?.allowMultiple, action?.allowedExtensions, onCompleteAction]);
+
+  const callback = useCallback((): void => {
+    try {
+      inputEl.click();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error(error);
+    }
+  }, [inputEl]);
+
+  return { callback };
+};
+
+export const useUploadFiles: EnsembleActionHook<UploadFilesAction> = (
+  action?: UploadFilesAction,
+) => {
+  const [body, setBody] = useState<Record<string, unknown>>();
+  const [headers, setHeaders] = useState<Record<string, unknown>>();
+  const [status, setStatus] = useState<UploadStatus>("pending");
+  const [progress, setProgress] = useState<number>(0.0);
+
+  const onCompleteAction = useEnsembleAction(action?.onComplete);
+  const onErrorAction = useEnsembleAction(action?.onError);
+  const screenContext = useScreenContext();
+
+  useRegisterBindings(
+    {
+      body,
+      headers,
+      status,
+      progress,
+    },
+    action?.id,
+    {
+      setBody,
+      setHeaders,
+      setStatus,
+      setProgress,
+    },
+  );
+
+  const apiModel = screenContext?.model?.apis?.find(
+    (model) => model?.name === action?.uploadApi,
+  );
+  if (!apiModel) return;
+
+  const progressCallback = (progressEvent: ProgressEvent): void => {
+    const percentCompleted = (progressEvent.loaded * 100) / progressEvent.total;
+
+    setProgress(percentCompleted);
+  };
+
+  const callback = async (): Promise<void> => {
```

Comment:
This function should be memoized. You can check for the matching API model and inputs within the callback and make it a no-op to avoid breaking rules of hooks. Likewise you can move progressCallback inside.